woocommerce/woocommerce
# Add handling for plugin-feature incompatibilities

### All Submissions:






### Changes proposed in this Pull Request:

After the features engine is completed (https://github.com/woocommerce/woocommerce/pull/34727) it's time to actually do something with the gathered compatibility info. This pull request implements UI changes in two places:

#### The feature settings page

If there are one or more _active_ plugins that haven't declared positive compatibility with a given feature, a warning notice will be displayed below the checkbox for the feature. Additionally, if the feature is disabled, the checkbox will be disabled too:

![image](https://user-images.githubusercontent.com/937723/194512662-c924d4fa-58c1-4470-923e-e34752538a1c.png)

The legacy features ("Analytics" and "New navigation") are excluded from this check and don't display any warning: these features have been around for a long time and it doesn't make much sense to start bugging plugin developers about declaring compatibility at this point.

#### The plugins page

The "Manage incompatible plugins" link in the feature notice targets a special view of the plugins page, with `?plugin_status=incompatible_with_feature&feature_id=<feature_id>`:

![image](https://user-images.githubusercontent.com/937723/194512845-a36df31f-5c3e-48a2-ba1b-e77bd460ce20.png)

This new view includes all the plugins that haven't declared positive compatibility with the given feature, regardless of whether the plugin is currently active or not. The link to the view will be visible only if `plugin_status=incompatible_with_feature` is present in the URL.

We are only showing the link to the `All` view together with our new view, because the counters would be wrong anyway (we are filtering the list returned by the `all_plugins` hook); the counter for `All` is explicitly recalculated here.

Regardless of the current view, for plugins that haven't declared positive compatibility with at least one of the _currently enabled_ features (again, excluding "Analytics" and "New navigation"), a warning will be shown:

![image](https://user-images.githubusercontent.com/937723/194531725-8df6c385-665b-4cef-8ea2-deedd3619159.png)

Additionally, the _Activate_ link for the plugin will be disabled.

#### Enabling/activating via code

It's still possible to activate plugins and enable features via code (`wp` cli tool) and that's on purpose: at least initially there must be a mechanism to bypass the UI constraints for the cases where the admins know what they are doing.

Additionally, the `FeatureUtil` class gets two new methods that disable the UI restrictions. So the admin can add this code to their site:

```php
add_action('before_woocommerce_init', function() {
	\Automattic\WooCommerce\Utilities\FeaturesUtil::allow_activating_plugins_with_incompatible_features();
	\Automattic\WooCommerce\Utilities\FeaturesUtil::allow_enabling_features_with_incompatible_plugins();
});
```

Then the warning notices will still be displayed, but it will be possible to activate the plugins/enable the features from the respective UI pages.

Closes #34861.

### How to test the changes in this Pull Request:

You may want to add or remove features during your testing. To do so, just manually modify the `$features` array that is initialized at tthe beginning of the constructor in the `FeaturesController` class.

#### Testting UI

You need to test that the features page (WooCommerce - Settings - Advanced - Features) and the plugins page work as expected by trying various combinations of enabling/disabling features and activating/deactivation plugins. Thinks to look for:

* Features can't be enabled from the settings page if there's at least one _active_ plugin that isn't compatible (but they can be disabled if they are already enabled).
* Plugins can't be activated from the plugins page if there's at least one _enabled_ feature that isn't compatible.
* The "Manage incompatible plugins" link redirects to a special view in the plugins page with only two views, "All" and "Incompatible with..."
* The special plugins view displays all the plugins that are incompatible with the feature, not only the active ones.
* The warning text for plugins only mentions/counts _enabled_ features, and the warning text for features only mentions/counts _active_ plugins.
* The text of the incompatibility warnings for both features and plugins renders appropriately for the cases of one, two and tthree or more incompatible items ("incompatible with X", "incompatible with X and Y", "incompatible with X, Y and N more")
* The plugin information rows renders nicely with all of these combinations:
  * Compatible and incompatible plugins
  * With and without compatibility warning notice
  * With and without plugin update available notice

#### Testing code

1. Add the snippet indicated above (in "Enabling/activating via code") and verify that the warnings are still displayed, but features can always be enabled and plugins can always be activated, even with incompatibilities.

2. Verify that you can still enable and disable features via code, even features with incompatibilities, and even without the above snippet, e.g: `wp option set woocommerce_feature_high_performance_order_storage_enabled yes`

3. Verify that you can still activate and deactivate plugins, even plugins with incompatibilities, and even without the above snippet, e.g: `wp plugin activate woocommerce-bookings`

### Other information:




### FOR PR REVIEWER ONLY:
